### ISSUE TITLE
[Documentation]: Documentation for `argTypes` is not correct

### ISSUE THREAD

**Issue by todor-a** [storybookjs/storybook#24174]
### Describe the problem

The docs say:

<img width="707" alt="image" src="https://github.com/storybookjs/storybook/assets/45210624/339e5381-f5d8-4759-9616-05e64ee43e33">

But the actual type is:

```
interface InputType {
    name?: string;
    description?: string;
    defaultValue?: any;
    type?: SBType | SBScalarType['name'];
    if?: Conditional;
    [key: string]: any;
}
```

I'm willing to work on this if it's indeed wrong.

**Comment by raphaelvdossantos** [comment]
Sorry, I am not sure I understand the issue here.
[The Docs](https://storybook.js.org/docs/react/essentials/controls#choosing-the-control-type) are describing the usage of a control type for an example of a button.
The `argTypes` variants are meant to restrict what options are possible to be selected for this property  in this use-case

As seen in the doc image right above this example, the user here can input any value in the `variant` field.
![image](https://github.com/storybookjs/storybook/assets/47623043/ee7bc8a2-b9b2-40e7-bdfa-94d2c52b399c)

The example illustrates a use-case where we want to restrict this text field to allow for only two options: `primary` or `secondary`. 

Could you please explain further what you mean by incorrect type?

**Comment by raphaelvdossantos** [comment]
I haven't checked the code, but the interface mentioned describes `[key: string]: any`.
This allows for dynamic extension of the object.

In this case, we might want, as per the docs, to generate a specific control for an unknown property.
By typing the interface `InputType` with  `[key: string]: any` we allow the user to extend the object dynamically, adding new properties to be controlled/restricted.

This is most likely intended and it is indeed documented.

Ex.: 
```javascript
const myArgs : InputType = {
// key fontStyle is of type string with value any and is permitted by the interface
 fontStyle: {
    options: ["bold", "italic", "regular"]
    }
 }
 ```

**Comment by todor-a** [comment]
Yes. I think adding options and control as optional but typed fields to the interface will be benfitial.

**Comment by shilman** [comment]
@kasperpeulen do you have any suggestions here?

**Comment by kylegach** [comment]
@kasperpeulen will likely have a better technical explanation (possibly with workarounds), but the situation here is basically:

argType properties like `options`, and `control` are contributed by the Controls addon and not part of the core argTypes API. That's why they're not in the types.

We agree this isn't ideal, from an end-user standpoint, and have some future plans to improve this, both in the types and in the documentation.

**Comment by noahgregory-basis** [comment]
@kylegach could/does the Controls addon at least expose its own types so we can at least have type safety?

**Comment by kylegach** [comment]
That's a good question. The source is here, https://github.com/storybookjs/storybook/tree/next/code/addons/controls, but I didn't see anything when I glanced around just now. Kasper will be able to weigh in more.

**Comment by noahgregory-basis** [comment]
The types from here appear to likely be good candidates for use in this:
https://github.com/storybookjs/storybook/blob/1013ca1/code/ui/blocks/src/controls/types.ts

Although they do appear to be only tangentially related to the Controls addon.

**Comment by jonniebigodes** [comment]
Closing this as the documentation was updated to contain the relevant APIs (including types). 

Will reopen if an update is required.
